src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -21,100 +20,77 @@ internal static class TokenOperations
-            );
-
-            // Find comments regions <# -> #>
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.Comment, RegionKindComment)
-            );
-
-            // Remove any null entries. Nulls appear if the folding reference is invalid
-            // or missing
-            foldableRegions.RemoveAll(item => item == null);
-
-            // Sort the FoldingReferences, starting at the top of the document,
-            // and ensure that, in the case of multiple ranges starting the same line,
-            // that the largest range (i.e. most number of lines spanned) is sorted
-            // first. This is needed to detect duplicate regions. The first in the list
-            // will be used and subsequent duplicates ignored.
-            foldableRegions.Sort();
-
-            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
-            // line number as the previous region. Therefore only emit ranges which have a different starting line
-            // than the previous range.
-            foldableRegions.RemoveAll( (FoldingReference item) => {
-                // Note - I'm not happy with searching here, but as the RemoveAll
-                // doesn't expose the index in the List, we need to calculate it. Fortunately the
-                // list is sorted at this point, so we can use BinarySearch.
-                int index = foldableRegions.BinarySearch(item);
-                if (index == 0) { return false; }
-                return (item.StartLine == foldableRegions[index - 1].StartLine);
-            });
+            for (int index = 0; index < tokens.Length; index++)
+            {
+                Token token = tokens[index];
+                if (token.Kind != TokenKind.Comment) { continue; }
+                // Processing for comment regions <# -> #>
+                if (token.Extent.StartLineNumber != token.Extent.EndLineNumber)
+                {
+                    FoldingReference foldRef = CreateFoldingReference(token, token, RegionKindComment);
+                    if (foldRef != null) { foldableRegions.Add(foldRef); }
+                    continue;
+                }
 
-            // Some editors have different folding UI, sometimes the lastline should be displayed
-            // If we do want to show the last line, just change the region to be one line less
-            if (ShowLastLine) {
-                foldableRegions.ForEach( item => { item.EndLine--; });
+                if (IsBlockComment(index, tokens))
+                {
+                    // Regex's are very expensive.  Use them sparingly!
+                    // Processing for # region -> # endregion
+                    if (s_startRegionTextRegex.IsMatch(token.Text))
+                    {
+                        tokenCommentRegionStack.Push(token);
+                        continue;
+                    }
```

Comment:
I would put a newline here and in places below where new condition blocks begin

Reply:
I'm not convinced on this.  I kept them grouped because  they are all to do with `# region` detection.  I didn't refactor to a method because I merged two functions together to create this.  Otherwise I end up doing too many regexes and comparions.